include/ppx/knob.h

```diff
@@ -0,0 +1,298 @@
+            PPX_ASSERT_MSG(!isspace(ch), "knob cannot be created as its name has whitespace");
+        }
+        PPX_ASSERT_MSG(mFlagNames.count(flagName) == 0, "knob with this name already exists");
+
+        T* knobPtr = new T(flagName, std::forward<ArgsT>(args)...);
+
+        RegisterKnob(flagName, knobPtr, parent);
+        return knobPtr;
+    }
+
+    // ImGui
+    void DrawAllKnobs(bool inExistingWindow = false);
+
+    // Command-line flags
+    std::string GetUsageMsg();
+    void        UpdateFromFlags(const CliOptions& opts);
+
+    // RegisterKnob will store newly created knob
+    void RegisterKnob(std::string flagName, Knob* newKnob, Knob* parent);
+};
+
+// Knob is an abstract class which contains common features for all knobs and knob hierarchy
+class Knob
+{
+    friend void KnobManager::RegisterKnob(std::string flagName, Knob* newKnob, Knob* parent);
+
+public:
+    Knob(const std::string& flagName)
+        : mFlagName(flagName), mDisplayName(flagName) {}
+
+    std::string        GetDisplayName() const { return mDisplayName; }
+    std::string        GetFlagName() const { return mFlagName; }
+    std::string        GetFlagHelp() const { return mFlagDesc; }
+    std::vector<Knob*> GetChildren() const { return mChildren; }
+
+    // Populate flag members
+    void SetDisplayName(std::string displayName) { mDisplayName = displayName; }
+    void SetFlagDesc(std::string flagDesc) { mFlagDesc = flagDesc; }
+
+    virtual void Draw() = 0;
+
+    // Flag description in usage message
+    virtual std::string GetFlagHelpText() const = 0;
+
+    // Updates knob value from commandline flag
+    virtual void UpdateFromFlag(const CliOptions& opts) = 0;
+
+    // Default values are the values when the application starts. The knobs can be reset to default by a button in the UI
+    virtual void ResetToDefault() = 0;
+
+private:
+    // Child knobs are displayed in the UI below their parent and slightly indented
+    void AddChild(Knob* pChild) { mChildren.push_back(pChild); }
+
+protected:
+    std::string        mDisplayName;
+    std::string        mFlagName;
+    std::string        mFlagDesc;
+    std::vector<Knob*> mChildren;
+};
+
+// KnobBoolCheckbox is a boolean knob that will be displayed as a checkbox in the UI
+class KnobBoolCheckbox
+    : public Knob
+{
+public:
+    KnobBoolCheckbox(const std::string& flagName, bool defaultValue, OnChangeCallback<bool> callback = nullptr)
+        : Knob(flagName)
+    {
+        mCallback = callback;
+        SetDefault(defaultValue);
+    }
+
+    void Draw() override;
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // GetFlagHelpText format:
+    // --flag_name <true|false> : flag description
+    std::string GetFlagHelpText() const override;
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <true|false>
+    void UpdateFromFlag(const CliOptions& opts) override;
+
+    bool GetValue() const { return mValue; }
```

Comment:
Seems like `GetValue` / `SetValue` could be pure virtual in the `Knob`* parent, and overriden by each knob, as it seems to be a common behavior.

*`Knob`, since the `KnobGroup` suggested in another comment would be the `KnobManager` you have today.

Reply:
The value type is different for each knob, i.e. int, float, Extend2D